logstash-core/lib/logstash/output_delegator.rb

```diff
@@ -8,26 +8,92 @@
 #
 # This plugin also records some basic statistics
 module LogStash; class OutputDelegator
-  attr_reader :workers, :config, :worker_count
+  attr_reader :workers, :config, :worker_count, :threadsafe
 
   # The *args this takes are the same format that a Outputs::Base takes. A list of hashes with parameters in them
   # Internally these just get merged together into a single hash
-  def initialize(logger, klass, *args)
+  def initialize(logger, klass, default_worker_count, *args)
     @logger = logger
+    @threadsafe = klass.threadsafe?
     @config = args.reduce({}, :merge)
     @klass = klass
-    @worker_count = @config["workers"] || 1
 
+    # We define this as an array regardless of threadsafety
+    # to make reporting simpler, even though a threadsafe plugin will just have
+    # a single instance
+    #
+    # Older plugins invoke the instance method Outputs::Base#workers_not_supported
+    # To detect these we need an instance to be created first :()
+    # TODO: In the next major version after 2.x remove support for this
+    @workers = [@klass.new(*args)]
+    @workers.first.register # Needed in case register calls `workers_not_supported`
+
+    # DO NOT move this statement before the instantiation of the first single instance
+    # Read the note above to understand why
+    @worker_count = calculate_worker_count(default_worker_count)
+    @logger.debug("Will start workers for output", :worker_count => @worker_count, :class => klass)
+
+    warn_on_worker_override!
+    # This queue is used to manage sharing across threads
     @worker_queue = SizedQueue.new(@worker_count)
 
-    @workers = @worker_count.times.map do
-      w = @klass.new(*args)
-      w.register
-      @worker_queue << w
-      w
+    @workers += (@worker_count - 1).times.map do
+      inst = @klass.new(*args)
+      inst.register
+      inst
     end
 
+    @workers.each { |w| @worker_queue << w }
+
     @events_received = Concurrent::AtomicFixnum.new(0)
+
+
+    # One might wonder why we don't use something like
+    # define_singleton_method(:multi_receive, method(:threadsafe_multi_receive)
+    # and the answer is this is buggy on Jruby 1.7.x . It works 98% of the time!
+    # The other 2% you get weird errors about rebinding to the same object
+    # Until we switch to Jruby 9.x keep the define_singleton_method parts
+    # the way they are, with a block
```

Comment:
what do you think about adding the link to the issue or at least the issue number?